std/json.zig

```diff
@@ -0,0 +1,1308 @@
+                    if (not_first) {
+                        std.debug.warn(",");
+                    }
+                    not_first = true;
+                    std.debug.warn("\"{}\":", entry.key);
+                    entry.value.toString();
+                }
+                std.debug.warn("}}");
+            },
+        }
+    }
+
+    pub fn toStringIndent(self: &const Value, indent: usize) void {
+        if (indent == 0) {
+            self.toString();
+        } else {
+            self.toStringIndentLevel(indent, 0);
+        }
+    }
+
+    fn toStringIndentLevel(self: &const Value, indent: usize, level: usize) void {
+        switch (*self) {
+            Value.Null => {
+                std.debug.warn("null");
+            },
+            Value.Bool => |inner| {
+                std.debug.warn("{}", inner);
+            },
+            Value.Integer => |inner| {
+                std.debug.warn("{}", inner);
+            },
+            Value.Float => |inner| {
+                std.debug.warn("{.5}", inner);
+            },
+            Value.String => |inner| {
+                std.debug.warn("\"{}\"", inner);
+            },
+            Value.Array => |inner| {
+                var not_first = false;
+                std.debug.warn("[\n");
+
+                for (inner.toSliceConst()) |value| {
+                    if (not_first) {
+                        std.debug.warn(",\n");
+                    }
+                    not_first = true;
+                    padSpace(level + indent);
+                    value.toStringIndentLevel(indent, level + indent);
+                }
+                std.debug.warn("\n");
+                padSpace(level);
+                std.debug.warn("]");
+            },
+            Value.Object => |inner| {
+                var not_first = false;
+                std.debug.warn("{{\n");
+                var it = inner.iterator();
+
+                while (it.next()) |entry| {
+                    if (not_first) {
+                        std.debug.warn(",\n");
+                    }
+                    not_first = true;
+                    padSpace(level + indent);
+                    std.debug.warn("\"{}\": ", entry.key);
+                    entry.value.toStringIndentLevel(indent, level + indent);
+                }
+                std.debug.warn("\n");
+                padSpace(level);
+                std.debug.warn("}}");
+            },
+        }
+    }
+
+    fn padSpace(indent: usize) void {
+        var i: usize = 0;
+        while (i < indent) : (i += 1) {
+            std.debug.warn(" ");
+        }
+    }
+};
+
+// A non-stream JSON parser which constructs a tree of Value's.
+const JsonParser = struct {
+    allocator: &Allocator,
+    state: State,
+    copy_strings: bool,
+    // Stores parent nodes and un-combined Values.
+    // Worst case scenario we have nested key, values and so need two times the stack size.
+    stack: [2 * StreamingJsonParser.max_stack_size]Value,
```

Comment:
max_stack_size should probably be a runtime parameter to `init` and then we would allocate stack on the heap. Because we may want to decide how much nesting to allow in the json data at runtime - or at least in the API of std.json

Reply:
I was thinking if it was worthwhile to make the `max_stack_size` parameter in the streaming parser a comptime variable for this specific case. For the value decoder, I'd actually change this to just use an ArrayList and not be dependent on the streaming parsers stack size. The rationale being that the common case will likely not be nested so reduce memory usage for a few extra dynamic allocations/copies in the worst case.